File: javabackend/src/main/java/mft/dev/javabackend/controller/UtenteController.java
package mft.dev.javabackend.controller;

import mft.dev.javabackend.dto.insert.request.UtenteInsertRequestDTO;
import mft.dev.javabackend.dto.insert.response.UtenteResponseDTO;
import mft.dev.javabackend.dto.update.UtenteUpdateRequestDTO;
import mft.dev.javabackend.model.Utente;
import mft.dev.javabackend.service.UtenteService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.NoSuchElementException;

@RestController
@RequestMapping("/utenti")
public class UtenteController {

    @Autowired
    private UtenteService utenteService;

    @GetMapping
    public ResponseEntity<List<UtenteResponseDTO>> getAllUtenti(){
        List<UtenteResponseDTO> utenti = this.utenteService.getUtenti().stream().map(UtenteResponseDTO::fromUtente).toList();

        return new ResponseEntity<>(utenti, HttpStatus.OK);
    }

    @GetMapping("/{id}")
    public ResponseEntity<UtenteResponseDTO> getUtente(@PathVariable("id") String id){
        long idConverted;

        try {
            idConverted = Long.parseLong(id);
        } catch (NumberFormatException ex){
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }

        try {
            Utente utenteFound = this.utenteService.getUtente(idConverted);

            return new ResponseEntity<>(UtenteResponseDTO.fromUtente(utenteFound), HttpStatus.OK);
        } catch (NoSuchElementException ex) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
    }

    @PostMapping
    public ResponseEntity<UtenteResponseDTO> insertUtente(@RequestBody UtenteInsertRequestDTO dto){
        Utente newUtente = this.utenteService.insertUtente(dto);

        return new ResponseEntity<>(UtenteResponseDTO.fromUtente(newUtente), HttpStatus.OK);
    }

    @PutMapping("/{id}")
    public ResponseEntity<UtenteResponseDTO> updateUtente(@PathVariable("id") String id, @RequestBody UtenteUpdateRequestDTO dto){
        long idConverted;

        try {
            idConverted = Long.parseLong(id);
        } catch (NumberFormatException ex){
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }

        try{
            Utente utenteUpdated = this.utenteService.updateUtente(idConverted, dto);

            return new ResponseEntity<>(UtenteResponseDTO.fromUtente(utenteUpdated), HttpStatus.OK);
        } catch (NoSuchElementException ex) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<String> deleteUtente(@PathVariable("id") String id){
        long idConverted;

        try {
            idConverted = Long.parseLong(id);
        } catch (NumberFormatException ex){
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }

        try {
            this.utenteService.deleteUtente(idConverted);

            return new ResponseEntity<>("Utente eliminato con successo!", HttpStatus.OK);
        } catch (NoSuchElementException ex) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
    }
}

File: kotlinbackend/src/main/kotlin/mft/dev/kotlinbackend/controller/UtenteController.kt
package mft.dev.kotlinbackend.controller

import mft.dev.kotlinbackend.dto.insert.UtenteInsertRequestDTO
import mft.dev.kotlinbackend.dto.response.TodoResponseDTO
import mft.dev.kotlinbackend.dto.response.UtenteResponseDTO
import mft.dev.kotlinbackend.dto.update.UtenteUpdateRequestDTO
import mft.dev.kotlinbackend.model.Todo
import mft.dev.kotlinbackend.model.Utente
import mft.dev.kotlinbackend.service.UtenteService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/utenti")
class UtenteController @Autowired constructor(
    val utenteService: UtenteService
) {

    @GetMapping
    fun getAllUtenti(): ResponseEntity<MutableList<UtenteResponseDTO>> {
        val utenti: MutableList<Utente> = this.utenteService.getUtenti()

        return ResponseEntity(utenti.stream().map { UtenteResponseDTO.fromUtente(it) }.toList(), HttpStatus.OK)
    }

    @GetMapping("/{id}")
    fun getUtente(@PathVariable("id") id: String): ResponseEntity<UtenteResponseDTO> {
        val idConverted: Long

        try {
            idConverted = id.toLong()
        } catch (ex: NumberFormatException) {
            return ResponseEntity(HttpStatus.BAD_REQUEST)
        }

        try {
            val utenteFound: Utente = this.utenteService.getUtente(idConverted)

            return ResponseEntity(UtenteResponseDTO.fromUtente(utenteFound), HttpStatus.OK)
        } catch (ex: NoSuchElementException) {
            return ResponseEntity(HttpStatus.NOT_FOUND)
        }

    }

    @PostMapping
    fun insertUtente(@RequestBody dto: UtenteInsertRequestDTO): ResponseEntity<UtenteResponseDTO> {
        val newUtente: Utente = this.utenteService.insertUtente(dto)

        return ResponseEntity(UtenteResponseDTO.fromUtente(newUtente), HttpStatus.OK)
    }

    @PutMapping("/{id}")
    fun updateUtente(@PathVariable("id") id: String, @RequestBody dto: UtenteUpdateRequestDTO): ResponseEntity<UtenteResponseDTO> {
        val idConverted: Long

        try {
            idConverted = id.toLong()
        } catch (ex: NumberFormatException) {
            return ResponseEntity(HttpStatus.BAD_REQUEST)
        }

        try {
            val utenteFound: Utente = this.utenteService.updateUtente(idConverted, dto)

            return ResponseEntity(UtenteResponseDTO.fromUtente(utenteFound), HttpStatus.OK)
        } catch (ex: NoSuchElementException) {
            return ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }

    @DeleteMapping("/{id}")
    fun deleteUtente(@PathVariable("id") id: String): ResponseEntity<String>{
        val idConverted: Long

        try {
            idConverted = id.toLong()
        } catch (ex: NumberFormatException) {
            return ResponseEntity(HttpStatus.BAD_REQUEST)
        }

        try {
            this.utenteService.deleteUtente(idConverted)

            return ResponseEntity("Utente eliminato con successo!", HttpStatus.OK)
        } catch (ex: NoSuchElementException) {
            return ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }
}